cpp/src/io/parquet/reader_impl_helpers.cpp

```diff
@@ -868,6 +884,16 @@ ColumnChunkMetaData const& aggregate_reader_metadata::get_column_metadata(size_t
                                                                           size_type src_idx,
                                                                           int schema_idx) const
 {
+  // schema_idx_maps will only have > 0 size when we are reading matching column projection from
+  // mismatched Parquet sources.
+  if (src_idx and schema_idx_maps.size()) {
+    auto const& schema_idx_map = schema_idx_maps[src_idx - 1];
+    CUDF_EXPECTS(schema_idx_map.find(schema_idx) != schema_idx_map.end(),
+                 "Unmapped schema index encountered in the specified source tree",
+                 std::out_of_range);
```

Comment:
Chose the error types for these exceptions such that they don't translate to `RuntimeError`s in Python while making the most sense. See this [thread](https://github.com/rapidsai/cudf/pull/16394#discussion_r1695578494) for more info. Please feel free to suggest if you think another exception would be appropriate

Reply:
I remember trying to throw `std::invalid_value` in C++ here and other places but it unfortunately translates to `RuntimeError` in Python. I will confirm this again.

Reply:
But even if so, why is a RuntimeError in Python _bad_? I don’t understand the limitations on the Python side. It seems like a RuntimeError is natural for this case.

Reply:
Here is the exception map: https://cython.readthedocs.io/en/latest/src/userguide/wrapping_CPlusPlus.html#exceptions

Reply:
>But even if so, why is a RuntimeError in Python bad? I don’t understand the limitations on the Python side. It seems like a RuntimeError is natural for this case.

I agree that RuntimeError here feels much more natural from, at least, C++'s perspective. Maybe @galipremsagar can elaborate on the Python side. 

>Here is the exception map: https://cython.readthedocs.io/en/latest/src/userguide/wrapping_CPlusPlus.html#exceptions

Wonderful. Thanks for pointing me to this. Let me use this to see if I can make things better/more descriptive.

Reply:
>Wonderful. Thanks for pointing me to this. Let me use this to see if I can make things better/more descriptive.

I have updated a couple of `out_of_range`s to `range_error`s but not sure if that helps much. Can't say a better (non-runtime exception) can be done for certain throws.